nservicebus/nhibernate/custom-mappings.md

```diff
@@ -0,0 +1,125 @@
+
+
+NOTE: The `[Unique]` attribute will not be used to generate a unique index. You are responsible to add this to your custom mapping. This is especially important when using both optimistic concurrency and a transaction isolation level different from serializable and allow concurrent processing of messages. Not adding this can result in multiple saga entities be persisted as the second insert will not fail when inserting a duplicate saga key (different from the primary key column) when multiple message  are processed concurrently that target the same saga instance.
+
+
+## Custom .hbm.xml mapping
+
+Using NHibernate mapping files is the original way to customize your mapping. It needs to be created and then either embedded as a resource in your assembly or available on the file system.
+
+Look at these pages to see how you can create these mappings:
+
+* http://nhibernate.info/doc/tutorials/first-nh-app/your-first-nhibernate-based-application.html
+
+
+### Mappings as embedded resources
+
+It is not required to create your own NHibernate configuration object and pass it to the NServiceBus NHibernate configuration.
+
+
+### Read mappings from file system instead of embedded resources
+
+If you do not want to embed the `*.hbm.xml` files then you can also load mappings from the file system. Create a custom NHibernate configuration object and use the following example to add mappings from the file system.
+
+<!-- import NHibernateLoadMappingsFromFileSystem -->
+
+## Use Fluent NHibernate
+
+Fluent NHibernate gives you a type-safe mapping approach where the mapping is specified as c# code and not as xml but still seperate from the classes. The benefit is that you get compile time feedback when a mapping is not valid any more when you have breaking changes in your mapping but also that your classes and mappings are not part of the same .net type.
+
+To use it with NServiceBus:
+
+1. Install `FluentNHibernate` package via Nuget.
+2. Create a custom NHibernate configuration
+a. via FluentNHibernate
+b. or by creating a new Configuration instance and pass it to FluentNHibernate
+3. Pass it to the NServiceBus NHibernate configuration.
+
+
+Example of a possible implementation:
+
+<!-- import NHibernateInitWithFluentNHibernate
+
+Its best to read the [Getting started](https://github.com/jagregory/fluent-nhibernate/wiki/Getting-started) article from the FluentNHibernate project to see how you create mappings.
+
+
+References:
+
+* http://www.fluentnhibernate.org
+* http://www.nuget.org/packages/FluentNHibernate/
+* https://github.com/jagregory/fluent-nhibernate
+* https://github.com/jagregory/fluent-nhibernate/wiki/Getting-started
+
+
+## Use NHibernate.Mapping.Attributes
+
+With NHibernate.Attributes you can decorate your saga data classes. This keeps your classes, mapping and schema data very close. Your saga types have a dependency on the NHibernate.Attributes assembly.
```

Comment:
the title is `NHibernate.Mapping.Attributes` but on this line it says  `NHibernate.Attributes`. which one is it?